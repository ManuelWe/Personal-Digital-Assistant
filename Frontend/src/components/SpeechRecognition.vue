<template>
  <div id="speechRecognition">
    <b-tooltip
      label="Mute"
      position="is-bottom"
      type="is-light"
    >
      <button
        v-if="listening"
        class="button is-info"
        @click="stopSpeechRecognition"
      >
        <span class="icon">
          <i class="fas fa-microphone fa-2x" />
        </span>
      </button>
    </b-tooltip>
    <b-tooltip
      label="Unmute"
      position="is-bottom"
      type="is-light"
    >
      <button
        v-if="!listening"
        class="button is-info"
        @click="startSpeechRecognition"
      >
        <span class="icon">
          <i class="fas fa-microphone-slash fa-2x" />
        </span>
      </button>
    </b-tooltip>
  </div>
</template>

<script>
export default {
  props: {
    userInput: {
      type: String,
      required: false,
      default: '',
    },
  },
  data() {
    return {
      listening: true,
      recognition: null,
      interimResult: '',
      recognizing: false,
      autoRestart: true,
      lastStartedAt: 0,
      isFinal: true,
      fixedUserInput: '',
    };
  },
  created() {
    this.initSpeechRecognition();
    this.startSpeechRecognition();
  },
  methods: {
    capitalize(s) {
      return s.replace(s.substr(0, 1), (m) => m.toUpperCase());
    },
    linebreak(s) {
      return s.replace(/\n\n/g, '<p></p>').replace(/\n/g, '<br>');
    },
    startSpeechRecognition() {
      this.listening = true;
      this.recognition.lang = 'en-US';
      this.recognition.start();
      this.interimResult = '';
      if (this.userInput !== '') {
        this.$emit('update:user-input', `${this.userInput} `);
      }
      this.lastStartedAt = new Date().getTime();
    },
    stopSpeechRecognition() {
      this.listening = false;
      this.autoRestart = false;
      if (this.recognizing) {
        this.recognition.stop();
      }
    },
    initSpeechRecognition() {
      let finalTranscript = '';
      window.SpeechRecognition = window.webkitSpeechRecognition || window.SpeechRecognition;
      this.recognition = new window.SpeechRecognition();

      this.recognition.continuous = true;
      this.recognition.interimResults = true;

      this.recognition.onstart = () => {
        this.recognizing = true;
      };

      this.recognition.onerror = (event) => {
        switch (event.error) {
          case 'not-allowed':
          case 'service-not-allowed':
          // if permission to use the mic is denied, turn off auto-restart
            this.autoRestart = false;
            console.error(event.error);
            break;
          default:
        }
      };

      this.recognition.onend = () => {
        this.recognizing = false;
        // auto restart if closed automatically and not by user action
        if (this.autoRestart) {
          // never restart automatically more than once per second
          const timeSinceLastStart = new Date().getTime() - this.lastStartedAt;
          this.autoRestartCount += 1;
          if (timeSinceLastStart < 1000) {
            setTimeout(() => {
              this.startSpeechRecognition();
            }, 1000 - timeSinceLastStart);
          } else {
            this.startSpeechRecognition();
          }
        }
      };

      this.recognition.onresult = (event) => {
        let interimTranscript = '';
        if (this.isFinal) {
          this.fixedUserInput = this.userInput;
          this.isFinal = false;
        }
        for (let i = event.resultIndex; i < event.results.length; i += 1) {
          if (event.results[i].isFinal) {
            finalTranscript += event.results[i][0].transcript;
          } else {
            interimTranscript += event.results[i][0].transcript;
          }
        }

        finalTranscript = this.capitalize(finalTranscript);
        this.interimResult = this.linebreak(interimTranscript);
        if (finalTranscript === '') {
          this.$emit('update:user-input', this.fixedUserInput + this.interimResult);
        } else {
          this.$emit('update:user-input', this.fixedUserInput + finalTranscript);
          this.isFinal = true;
          finalTranscript = '';
        }
      };
    },
  },
};

</script>
