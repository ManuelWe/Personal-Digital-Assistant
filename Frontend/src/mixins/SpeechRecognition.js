/* eslint-disable new-cap */
export default {
  data() {
    return {
      finalSpan: '',
      interimSpan: '',
      finalTranscript: '',
      recognizing: false,
    };
  },
  created() {
    // eslint-disable-next-line no-undef
    const speechRecognition = new SpeechRecognition();
    speechRecognition.onresult = console.log;
    speechRecognition.start();
  },
  methods: {
    capitalize(s) {
      return s.replace(s.substr(0, 1), (m) => m.toUpperCase());
    },
    linebreak(s) {
      return s.replace(/\n\n/g, '<p></p>').replace(/\n/g, '<br>');
    },
    startDictation() {
      if (this.recognizing) {
        this.recognition.stop();
        return;
      }
      this.finalTranscript = '';
      this.recognition.lang = 'en-US';
      this.recognition.start();
      this.finalSpan.innerHTML = '';
      this.interimSpan.innerHTML = '';
    },
    test() {
      if ('webkitSpeechRecognition' in window) {
        // eslint-disable-next-line no-undef
        const recognition = new webkitSpeechRecognition();

        recognition.continuous = true;
        recognition.interimResults = true;

        recognition.onstart = () => {
          this.recognizing = true;
        };

        recognition.onerror = (event) => {
          console.log(event.error);
        };

        recognition.onend = () => {
          this.recognizing = false;
        };

        recognition.onresult = (event) => {
          let interimTranscript = '';
          for (let i = event.resultIndex; i < event.results.length; i += 1) {
            if (event.results[i].isFinal) {
              this.finalTranscript += event.results[i][0].transcript;
            } else {
              interimTranscript += event.results[i][0].transcript;
            }
          }
          this.finalTranscript = this.capitalize(this.finalTranscript);
          this.finalSpan.innerHTML = this.linebreak(this.finalTranscript);
          this.interimSpan.innerHTML = this.linebreak(interimTranscript);
        };
      }
    },
  },
};
