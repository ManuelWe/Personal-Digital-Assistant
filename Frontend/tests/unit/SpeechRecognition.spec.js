import { createLocalVue, mount } from '@vue/test-utils';
import Buefy from 'buefy';
import VueRouter from 'vue-router';
import SpeechRecognition from '@/components/SpeechRecognition.vue';

const localVue = createLocalVue();
localVue.use(Buefy);
localVue.use(VueRouter);

const routes = [
  {
    path: '/',
    component: SpeechRecognition,
  },
];
const router = new VueRouter({
  routes,
});

function factory() {
  return mount(SpeechRecognition, {
    localVue,
    router,
  });
}


describe('SpeechRecognition.vue', () => {
  describe('Elements are rendered correctly', () => {
    it('Component renders', () => {
      const wrapper = factory();
      expect(wrapper).toMatchSnapshot();
    });

    it('mute button renders, if listening', () => {
      const wrapper = factory();
      wrapper.vm.listening = true;
      wrapper.vm.$nextTick(() => {
        expect(wrapper.find('#muteButton').isVisible()).toBe(true);
      });
    });

    it('unmute button renders, if !listening', () => {
      const wrapper = factory();
      wrapper.vm.listening = false;
      wrapper.vm.$nextTick(() => {
        expect(wrapper.find('#unmuteButton').isVisible()).toBe(true);
      });
    });
  });

  describe('methods are working', () => {
    it('capitalize method capitalizes', () => {
      const wrapper = factory();
      expect(wrapper.vm.capitalize('test')).toBe('Test');
    });

    it('speech recognition is stoped', () => {
      const wrapper = factory();
      const spy = jest.spyOn(wrapper.vm, 'stopSpeechRecognition');
      wrapper.vm.stopSpeechRecognition();
      wrapper.vm.$nextTick(() => {
        expect(spy).toHaveBeenCalled();
      });
    });

    it('userInput is emitted if not empty', () => {
      const wrapper = factory();
      wrapper.setProps({ userInput: 'Hello Gunter' });
      wrapper.vm.startSpeechRecognition();
      expect(wrapper.emitted('update:user-input')[0]).toEqual(['Hello Gunter ']);
    });
  });

  describe('speech recognition action listeners working', () => {
    it('recognition.onend working', () => {
      const wrapper = factory();
      wrapper.vm.recognition.onend();
      expect(wrapper.vm.recognizing).toBe(false);
    });

    it('recognition.onerror working', () => {
      const wrapper = factory();
      wrapper.vm.recognition.onerror({ error: 'service-not-allowed' });
      expect(wrapper.vm.autoRestart).toBe(false);
    });

    describe('recognition.onresult working', () => {
      it('end recognition with interim result', () => {
        const wrapper = factory();
        wrapper.vm.recognition.onresult({
          resultIndex: 0,
          results: { length: 1, 0: { isFinal: false, 0: { transcript: 'Test' } } },
        });
        expect(wrapper.vm.interimResult).toBe('Test');
      });

      it('end recognition with final transcript', () => {
        const wrapper = factory();
        wrapper.vm.recognition.onresult({
          resultIndex: 0,
          results: { length: 1, 0: { isFinal: true, 0: { transcript: 'Test' } } },
        });
        expect(wrapper.vm.isFinal).toBe(true);
      });
    });

    it('recognition.onstart working', () => {
      const wrapper = factory();
      wrapper.vm.recognition.onstart();
      expect(wrapper.vm.recognizing).toBe(true);
    });
  });
});
