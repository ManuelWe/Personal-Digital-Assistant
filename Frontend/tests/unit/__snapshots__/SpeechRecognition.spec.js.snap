// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpeechRecognition.vue Elements are rendered correctly Component renders 1`] = `<div id="speechRecognition"><span data-label="Mute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"><button id="muteButton" class="button is-info"><span class="icon"><i class="fas fa-microphone fa-2x"></i></span></button></span> <span data-label="Unmute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"></span></div>`;
